src/util/Cache.h

```diff
@@ -227,54 +153,113 @@ class FlexibleCache {
+    auto ptr = make_shared<Value>(std::move(value));
+    return insertPinned(key, std::move(ptr));
+  }
+
+  /// Insert a key-value pair to the cache. Throws an exception if the key is
+  /// already present. Overload for shared_ptr of value. If the value is too big
+  /// for the cache, nothing happens.
+  ValuePtr insert(const Key& key, shared_ptr<Value> valPtr) {
+    if (contains(key)) {
+      throw std::runtime_error(
+          "Trying to insert a cache key which was already present");
+    }
+
+    // ignore elements that are too big
+    if (_valueSizeGetter(*valPtr) > _maxSizeSingleValue) {
+      return {};
+    }
+    Score s = _scoreCalculator(*valPtr);
+    _totalSizeNonPinned += _valueSizeGetter(*valPtr);
+    auto handle = _entries.insert(std::move(s), Entry(key, std::move(valPtr)));
     _accessMap[key] = handle;
-    // todo<joka921> : codeDuplication
-    if (_data.size() > _capacity) {
-      // Remove the last element.
-      // Since we are using shared_ptr this does not free the underlying
-      // memory if it is still accessible through a previously returned
-      // shared_ptr
-
-      _accessMap.erase(_data.pop().value().key());
+    shrinkToFit();
+    // The first value is the value part of the key-value pair in the priority
+    // queue (where the key is a score and the value is a cache entry). The
+    // second value is the value part of the key-value pair in the cache (where
+    // the key is the cache key and the value is the payload = in our case, a
+    // result).
+    return handle.value().value();
+  }
+
+  /// Insert a pinned key-value pair to the cache. Throws an exception if the
+  /// key is already present. Overload for shared_ptr<Value>. If the value is
+  /// too big for the cache, an exception is thrown.
+  ValuePtr insertPinned(const Key& key, shared_ptr<Value> valPtr) {
+    if (contains(key)) {
+      throw std::runtime_error(
+          "Trying to insert a cache key which was already present");
+    }
+
+    // throw if an element is too big for this cache
+    if (_valueSizeGetter(*valPtr) > _maxSizeSingleValue) {
+      throw std::runtime_error(
+          "Trying to pin an element to the cache that is bigger than the "
+          "maximum size for a single element in the cache");
     }
-    assert(_data.size() <= _capacity);
+    _pinnedMap[key] = valPtr;
+    _totalSizePinned += _valueSizeGetter(*valPtr);
+    shrinkToFit();
+    return valPtr;
   }
 
   //! Set the capacity.
-  void setCapacity(const size_t nofElements) {
-    std::lock_guard<std::mutex> lock(_lock);
-    _capacity = nofElements;
-    while (_data.size() > _capacity) {
-      // Remove elements from the back until we meet the capacity requirement
-      // TODO<joka921> : code duplication
-      _accessMap.erase(_data.pop().value().key());
-    }
-    assert(_data.size() <= _capacity);
+  void setCapacity(const size_t numElements) {
+    _capacityNumElements = numElements;
+    shrinkToFit();
```

Comment:
Elements or entries? I vote for entry (see other comments), but either is fine, but it should be consistent.

Also, we speak of "maximum number" in other places, so better call this setMaxNumEntries or setMaxNumElements